pkg/controllers/flannelmigration/ipam_migrator.go

```diff
@@ -0,0 +1,375 @@
+				log.WithError(err).Errorf("Failed to get existing default IPv4 IP pool")
+				return err
+			}
+
+			// Check if existing ip pool is valid.
+			if defaultPool.Spec.CIDR != cidr.String() ||
+				defaultPool.Spec.BlockSize != blockSize ||
+				defaultPool.Spec.NATOutgoing != isNATOutgoingEnabled ||
+				defaultPool.Spec.VXLANMode != api.VXLANModeAlways {
+				log.Errorf("Failed to validate existing default IPv4 IP pool %+v", defaultPool.Spec)
+				return cerrors.ErrorValidation{
+					ErroredFields: []cerrors.ErroredField{{
+						Name:   "pool.Spec",
+						Reason: "Failed to validate existing default IPv4 IP pool",
+					}},
+				}
+			}
+
+			log.Info("Use existing default IPv4 pool.")
+		}
+	} else {
+		log.Info("Created default IPv4 pool.")
+	}
+
+	return nil
+}
+
+// Update default FelixConfiguration with specified VNI, port and MTU.
+// Do nothing if correct values already been set.
+// Return error if default FelixConfiguration not exists or vxlan is not enabled.
+func updateDefaultFelixConfigurtion(ctx context.Context, client client.Interface, vni, port, mtu int) error {
+	// Get default Felix configuration. Return error if not exists.
+	defaultConfig, err := client.FelixConfigurations().Get(ctx, defaultFelixConfigurationName, options.GetOptions{})
+	if err != nil {
+		log.WithError(err).Errorf("Error getting default FelixConfiguration resource")
+		return err
+	}
+
+	// Check if vxlan is enabled. Return error is not.
+	vxlanEnabled := false
+	if defaultConfig.Spec.VXLANEnabled != nil {
+		vxlanEnabled = *defaultConfig.Spec.VXLANEnabled
+	}
+	if !vxlanEnabled {
+		log.WithError(err).Errorf("vxlan is not enabled by default Felix configration")
+		return err
+	}
+
+	// Get current value for VNI , Port and MTU.
+	currentVNI := 0
+	if defaultConfig.Spec.VXLANVNI != nil {
+		currentVNI = *defaultConfig.Spec.VXLANVNI
+	}
+	currentPort := 0
+	if defaultConfig.Spec.VXLANPort != nil {
+		currentPort = *defaultConfig.Spec.VXLANPort
+	}
+	currentMTU := 0
+	if defaultConfig.Spec.VXLANMTU != nil {
+		currentMTU = *defaultConfig.Spec.VXLANMTU
+	}
+
+	// Do nothing if the correct value has been set.
+	if currentVNI == vni && currentPort == port && currentMTU == mtu {
+		log.Infof("Default Felix configration got correct VNI(%d), port(%d), mtu(%d).", currentVNI, currentPort, currentMTU)
+		return nil
+	}
+
+	defaultConfig.Spec.VXLANVNI = &vni
+	defaultConfig.Spec.VXLANPort = &port
+	defaultConfig.Spec.VXLANMTU = &mtu
+	defaultConfig.Spec.InterfaceExclude = "kube-ipvs0,cni0" // prevent autodetecting cni0 interface.
+	_, err = client.FelixConfigurations().Update(ctx, defaultConfig, options.SetOptions{})
+	if err != nil {
+		// Migration controller should be the single source of updating FelixConfiguration.
+		// If an update conflict occur, it means there is a race between two migration controllers or
+		// between migration controller and a calico node. In both cases, something is wrong.
+		if _, ok := err.(cerrors.ErrorResourceUpdateConflict); ok {
```

Comment:
We're doing the same thing here independent of the type of error - do we really need separate blocks?

What happens when an error is hit, does something higher up the stack retry?